// react-date-picker-provider/lib/bar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DateBar;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _calendar = require('./calendar');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @param {Function} onChange 当前月份改变事件
 * @param {Date} month 当前月份
 * @param {Date} max 时间最大边界
 * @param {Date} min 时间最小边界 
 * @param {String} className 盒子样式
 */
function DateBar(_ref) {
  var onChange = _ref.onChange,
      month = _ref.month,
      max = _ref.max,
      min = _ref.min,
      className = _ref.className;

  min = (0, _calendar.getMaxDate)(min);
  max = (0, _calendar.getMinDate)(max);

  var _hasSwitchMonth = (0, _calendar.hasSwitchMonth)(month, max, min),
      isLastMonth = _hasSwitchMonth.isLastMonth,
      isNextMonth = _hasSwitchMonth.isNextMonth;

  return _react2.default.createElement(
    'div',
    { className: 'date-picker-bar-box ' + className },
    _react2.default.createElement(
      'button',
      {
        className: 'date-picker-bar-icon',
        disabled: !isLastMonth ? 'disabled' : '',
        onClick: function onClick(e) {
          onChange((0, _calendar.switchMonth)(month, -1));
        } },
      '\u4E0A\u4E2A\u6708'
    ),
    _react2.default.createElement(
      'div',
      { className: 'date-picker-bar-txt' },
      month.getFullYear() + '\u5E74' + (month.getMonth() + 1) + '\u6708'
    ),
    _react2.default.createElement(
      'button',
      {
        className: 'date-picker-bar-icon',
        disabled: !isNextMonth ? 'disabled' : '',
        onClick: function onClick() {
          onChange((0, _calendar.switchMonth)(month, 1));
        } },
      '\u4E0B\u4E2A\u6708'
    )
  );
}
DateBar.propTypes = {
  onChange: _propTypes2.default.func,
  month: _propTypes2.default.instanceOf(Date).isRequired,
  isNextMonth: _propTypes2.default.bool,
  isLastMonth: _propTypes2.default.bool,
  className: _propTypes2.default.string
};
DateBar.defaultProps = {
  onChange: _calendar.noop,
  isNextMonth: false,
  isLastMonth: false,
  min: '',
  max: '',
  className: ''
};
