// react-date-picker-provider/lib/itemBox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = DateBox;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _item = require('./item');

var _item2 = _interopRequireDefault(_item);

var _calendar = require('./calendar');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function DateBox(_ref) {
  var month = _ref.month,
      date = _ref.date,
      start = _ref.start,
      formatWeek = _ref.formatWeek,
      DateItem = _ref.DateItem,
      className = _ref.className,
      dateClassName = _ref.dateClassName,
      otherProps = _objectWithoutProperties(_ref, ['month', 'date', 'start', 'formatWeek', 'DateItem', 'className', 'dateClassName']);

  var monthDateList = (0, _calendar.getDisplayDates)(month, start, formatWeek);

  return _react2.default.createElement(
    'table',
    { className: className + ' date-picker-panel' },
    _react2.default.createElement(
      'tbody',
      null,
      monthDateList.map(function (item, index) {
        return _react2.default.createElement(
          'tr',
          { key: index },
          item.map(function (i) {
            return _react2.default.createElement(
              'td',
              { key: i.date },
              _react2.default.createElement(DateItem, _extends({}, i, {
                selectDate: (0, _calendar.toDateWithNoDiscrepancy)(date),
                className: dateClassName
              }, otherProps))
            );
          })
        );
      })
    )
  );
}
DateBox.propTypes = {
  onSelect: _propTypes2.default.func,
  className: _propTypes2.default.string,
  dateClassName: _propTypes2.default.string
};
DateBox.defaultProps = {
  onSelect: _calendar.noop,
  className: '',
  dateClassName: '',
  DateItem: _item2.default
};
