// react-date-picker-provider/lib/item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DateItem;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _calendar = require('./calendar');

var _config = require('./config');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @param {Date} min  最小日期边界
 * @param {Date} max  最大日期边界
 * @param {Date} selectDate 已选日期
 * @param {Date} date  当前日期
 * @param {Boolean} isCurMonth 是否是当前月份
 * @param {Number} dateNum  当月第几天
 * @param {Function} onSelect 选中事件
 * @param {String} className 外层样式
 */
function DateItem(_ref) {
  var min = _ref.min,
      max = _ref.max,
      isCurMonth = _ref.isCurMonth,
      selectDate = _ref.selectDate,
      dateNum = _ref.dateNum,
      onSelect = _ref.onSelect,
      date = _ref.date,
      className = _ref.className;

  min = (0, _calendar.getMaxDate)(min);
  max = (0, _calendar.getMinDate)(max);
  var isSelect = (0, _calendar.isDateEqual)(selectDate, date);
  var maxType = max ? (0, _calendar.getCompareDate)(date, max) : 1; // 0 : date === max ; 1: date < max ; 2 : date > max
  var minType = min ? (0, _calendar.getCompareDate)(date, min) : 2; // 0 : date === min ; 1: date < min ; 2 : date > min
  var isDisabled = maxType === 2 || minType === 1;
  var isToday = (0, _calendar.isDateEqual)(new Date(), date);
  return isCurMonth ? _react2.default.createElement(
    'button',
    {
      className: (isSelect ? 'select' : '') + ' date-picker-item-box ' + className,
      disabled: isDisabled ? 'disabled' : '',
      onClick: function onClick() {
        onSelect(date);
      } },
    isToday ? '今' : dateNum
  ) : _react2.default.createElement('div', null);
}
DateItem.propTypes = {
  onSelect: _propTypes2.default.func,
  isCurMonth: _propTypes2.default.bool.isRequired,
  className: _propTypes2.default.string
};
DateItem.defaultProps = {
  onSelect: _config.noop,
  className: '',
  min: '',
  max: '',
  selectDate: ''
};
